OpenRA.Game/Activities/Activity.cs

```diff
@@ -31,45 +31,56 @@ public abstract class Activity
 	{
 		public ActivityState State { get; private set; }
 
-		Activity childActivity;
-		protected Activity ChildActivity
-		{
-			get { return childActivity != null && childActivity.State < ActivityState.Done ? childActivity : null; }
-			set { childActivity = value; }
-		}
-
+		protected Activity ChildActivity { get; private set; }
 		public Activity NextActivity { get; protected set; }
 
 		public bool IsInterruptible { get; protected set; }
+		public bool ChildHasPriority { get; protected set; }
 		public bool IsCanceling { get { return State == ActivityState.Canceling; } }
 
 		public Activity()
 		{
 			IsInterruptible = true;
+			ChildHasPriority = true;
 		}
 
 		public Activity TickOuter(Actor self)
 		{
-			if (State == ActivityState.Done && Game.Settings.Debug.StrictActivityChecking)
-				throw new InvalidOperationException("Actor {0} attempted to tick activity {1} after it had already completed.".F(self, this.GetType()));
-
 			if (State == ActivityState.Queued)
 			{
 				OnFirstRun(self);
 				State = ActivityState.Active;
 			}
 
+			if (ChildHasPriority)
+			{
+				ChildActivity = ActivityUtils.RunActivity(self, ChildActivity);
+				if (ChildActivity != null)
+					return this;
+			}
+
 			var ret = Tick(self);
-			if (ret != this)
+
+			if (ChildActivity != null && (ChildActivity.State == ActivityState.Queued || !ChildHasPriority))
 			{
-				State = ActivityState.Done;
-				OnLastRun(self);
+				ChildActivity = ActivityUtils.RunActivity(self, ChildActivity);
+				if (!ChildHasPriority && ChildActivity == null)
```

Comment:
Can you please explain this? It doesn't make sense to me that a child activity completing should force the parent (which may still have work to do?) to be skipped without cancellation.

Reply:
`ChildHasPriority = false` is meant for activities that need to calculate stuff before the childactivity ticks _as long as there are children running_: Only `MoveAdjacentTo` and `AttackMoveActivity` need this. The other activities are happy to let the child finish before doing their own logic.

Perhaps it would be clearer to give the flag another name (although I'm not sure what that should be)?

Reply:
This feels a bit too magic for me - I can see us running into a brick wall if we find the need for a third type of activity in the future. I'm all for the current behavior with `ChildHasPriority = true`, but I'm still thinking that it would be simpler and more robust to make this flag completely disable the child `RunActivity` when disabled.  We could then expose a `TickChild` method that activities like `AttackMoveActivity` can call when when better suits them (and throws an exception if it is called by a method that hasn't opted out of the automatic behaviour). This would make the behaviour much more obvious.

Reply:
OK. `AttackMoveActivity` and `MoveAdjacentTo` now call `TickChild` themselves.